sqlx-postgres/src/types/geometry/path.rs

```diff
@@ -0,0 +1,334 @@
+#[derive(Debug, Clone, PartialEq)]
+pub struct PgPath {
+    pub closed: bool,
+    pub points: Vec<PgPoint>,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+struct Header {
+    is_closed: bool,
+    length: usize,
+}
+
+impl Type<Postgres> for PgPath {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("path")
+    }
+}
+
+impl PgHasArrayType for PgPath {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("_path")
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for PgPath {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        match value.format() {
+            PgValueFormat::Text => Ok(PgPath::from_str(value.as_str()?)?),
+            PgValueFormat::Binary => Ok(PgPath::from_bytes(value.as_bytes()?)?),
+        }
+    }
+}
+
+impl<'q> Encode<'q, Postgres> for PgPath {
+    fn produces(&self) -> Option<PgTypeInfo> {
+        Some(PgTypeInfo::with_name("path"))
+    }
+
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> Result<IsNull, BoxDynError> {
+        self.serialize(buf)?;
+        Ok(IsNull::No)
+    }
+}
+
+impl FromStr for PgPath {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let closed = !s.contains("[");
+        let sanitised = s.replace(&['(', ')', '[', ']', ' '][..], "");
+        let mut parts = sanitised.split(",");
+
+        let mut points = vec![];
+
+        while let (Some(x_str), Some(y_str)) = (parts.next(), parts.next()) {
+            let x = parse_float_from_str(x_str, "could not get x")?;
+            let y = parse_float_from_str(y_str, "could not get y")?;
+
+            let point = PgPoint { x, y };
+            points.push(point);
+        }
+
+        if !points.is_empty() {
+            return Ok(PgPath { points, closed });
+        }
+
+        Err(Error::Decode(
+            format!("could not get path from {}", s).into(),
+        ))
+    }
+}
+
+impl PgPath {
+    fn header(&self) -> Header {
+        Header {
+            is_closed: self.closed,
+            length: self.points.len(),
+        }
+    }
+
+    fn from_bytes(mut bytes: &[u8]) -> Result<Self, Error> {
+        let header = Header::try_read(&mut bytes).map_err(|s| Error::Decode(s.into()))?;
+
+        if bytes.len() != header.data_size() {
+            return Err(Error::Decode(
+                format!(
+                    "expected {} bytes after header, got {}",
+                    header.data_size(),
+                    bytes.len()
+                )
+                .into(),
+            ));
+        }
+
+        if bytes.len() % BYTE_WIDTH * 2 != 0 {
+            return Err(Error::Decode(
+                format!(
+                    "data length not divisible by pairs of {BYTE_WIDTH}: {}",
+                    bytes.len()
+                )
+                .into(),
+            ));
+        }
+
+        let mut out_points = Vec::with_capacity(bytes.len() / BYTE_WIDTH * 2);
```

Comment:
Are you sure that this capacity is not 4x what it should be?

Your current version is equivalent to

```rust
        let mut out_points = Vec::with_capacity(bytes.len() / size_of(f32));
```

Reply:
You're right, this is some order of operations silliness on my part. Have corrected, ty